File: solutions/javascript/0077-combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    const result = [];
    generateCombinations(1, n, k, [], result);
    return result;
};

function generateCombinations(start, n, k, combination, result) {
    if (k === 0) {
        result.push([...combination]);
        return;
    }
    for (let i = start; i <= n; ++i) {
        combination.push(i);
        generateCombinations(i + 1, n, k - 1, combination, result);
        combination.pop();
    }
}

File: solutions/java/0077-combinations.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        generateCombinations(1, n, k, new ArrayList<Integer>(), result);
        return result;
    }

    private void generateCombinations(int start, int n, int k, List<Integer> combination, List<List<Integer>> result) {
        if (k == 0) {
            result.add(new ArrayList<>(combination));
            return;
        }
        for (int i = start; i <= n - k + 1; i++) {
            combination.add(i);
            generateCombinations(i + 1, n, k - 1, combination, result);
            combination.remove(combination.size() - 1);
        }
    }
}
